Copy-On-Write for Key and KeySet

atmaxinger opened the pull request:
<!--
Check relevant points but **please do not remove entries**.
-->

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you,
but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For docu fixes, spell checking, and similar none of these points below
need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

atmaxinger commented:
This still has some rough edges, I created a PR mainly to see whether the CI builds pass, because locally I can't execute all the tests.

atmaxinger commented on src/plugins/xmltool/stream.c:
```diff
@@ -158,14 +158,14 @@ ssize_t keyToStreamBasename (const Key * key, FILE * stream, const char * parent
 		size_t skip = parentSize ? parentSize : elektraStrLen (parent) - 1;
 
-		found = memcmp (parent, key->key, skip);
+		found = memcmp (parent, keyName (key), skip);
```
Is there any reason why the XML plugin used internal fields of the struct directly instead of the public API?

markus2330 replied:
The plugin is older than most parts of the public API.

atmaxinger replied:
Well then it's good somebody finally got to fix it 😉

atmaxinger commented on src/plugins/mmapstorage/CMakeLists.txt:
```diff
@@ -1,3 +1,4 @@
+#[[
```
I've disabled the build of `mmapstorage` for now, because obviously the plugin does not compile at this time.

atmaxinger commented on src/plugins/hosts/keymetaformatting.c:
```diff
@@ -48,7 +48,9 @@ ssize_t keySetStringF (Key * key, const char * format, ...)
-	return key->dataSize;
+	return key->dataSize;*/
+
+	return keySetRaw (key, p, elektraStrLen (p));
```
Is there any reason why `keySetMeta` previously did not use `keySetRaw` to set the value, and instead reimplented part of the logic?

atmaxinger commented on src/libs/elektra/keymeta.c:
```diff
@@ -483,19 +482,7 @@ ssize_t keySetMeta (Key * key, const char * metaName, const char * newMetaString
-		clear_bit (toSet->flags, (keyflag_t) KEY_FLAG_MMAP_DATA);
-		toSet->data.c = metaStringDup;
-		toSet->dataSize = metaStringSize;
+		keySetRaw (toSet, newMetaString, metaStringSize);
```
Is there any reason why `keySetMeta` duplicated parts of the functionality of `keySetRaw` here?

markus2330 replied:
Nice catch! :heart: I don't think so.

kodebach replied:
I guess it was because of the `TODO`, or maybe this was simply implemented before `keySetRaw` existed.

markus2330 commented:
> This still has some rough edges, I created a PR mainly to see whether the CI builds pass, because locally I can't execute all the tests.

No need for explanations, always happy to see PRs! :revolving_hearts:

mpranj commented:
> No need for explanations, always happy to see PRs! 

Fully agree. In addition, sometimes it is easier to understand and discuss with real code. 👍 
Looking forward to review this.

kodebach commented:
> > No need for explanations, always happy to see PRs!
> 
> Fully agree. In addition, sometimes it is easier to understand and discuss with real code. :+1: Looking forward to review this.

Totally agree as well, although it technically goes against the review process, because it creates a bias. Though in this case, we all already agree that this is the best approach, so here it's fine.

kodebach commented on src/libs/elektra/key.c:
```diff
@@ -341,125 +341,85 @@ Key * keyCopy (Key * dest, const Key * source, elektraCopyFlags flags)
+	keyDataDel (dest->keyData, !test_bit (dest->flags, KEY_FLAG_MMAP_DATA));
 	ksDel (dest->meta);
 
-	*dest = orig;
```
This line was crucial... Together with having all the `elektraFree` _after_ the last `goto memerror`, it meant that `dest` would not be modified, if `keyCopy` fails.

The current version will corrupt `dest` on failure.

atmaxinger replied:
You're right. I've added the functionality back in and also created a test that checks exactly that.

kodebach commented on src/libs/elektra/keyname.c:
```diff
@@ -165,6 +165,52 @@
+ * @param source the keyname to copy
+ * @return copied keyname
+ */
+struct _KeyName * keyNameCopy (struct _KeyName * source)
```
```suggestion
static struct _KeyName * keyNameCopy (struct _KeyName * source)
```

atmaxinger replied:
No, the `keyNameCopy` function is used outside `keyname.c` aswell.

kodebach replied:
In that case, I'd move the `keyNameIncRef` into this function, because AFAICT, you always used `keyNameDecAndDel`, so having refcount zero would just be wrong.

kodebach commented on src/libs/elektra/keyname.c:
```diff
@@ -165,6 +165,52 @@
 #include "kdbinternal.h"
 
+/**
+ * Helper method: creates a deep copy of a keyname. Does not copy the refs counter.
```
```suggestion
 * Helper method: creates a deep copy of a keyname. Does not copy the refs counter, instead sets it to 0.
```

kodebach commented on src/libs/elektra/keyname.c:
```diff
@@ -255,10 +301,11 @@ keyDel(key);
-	if (!key) return 0;
+	if (!key || !key->keyName) return 0;
 
-	ELEKTRA_ASSERT (key->key != NULL, "invalid name");
-	return key->key;
+	ELEKTRA_ASSERT (key->keyName->key != NULL, "invalid name");
```
To match unescaped name below:

```suggestion
	if (!key) return 0;
	ELEKTRA_ASSERT (key->keyName != NULL, "invalid name");
	ELEKTRA_ASSERT (key->keyName->key != NULL, "invalid name");
```

atmaxinger replied:
Hmm, based on the docs: `@retval "" when Key's name is empty`.

So wouldn't it be better if we return an empty string if `key->keyName` is `NULL`?

markus2330 replied:
If you are talking about the return values of `keyName()`: Please be very faithful to the public API **exactly** as it is. The API was recently completely reviewed. Changes must be very careful, usually even with decisions and also all changes completely listed in the release notes (for later 1.0 migration documents). So it is not a good idea to do such changes within other PRs.

What `key->keyName` stores is more or less an implementation detail. It is actually going to change, together with an API change of `keyName` which probably will solve anyway what you ask here? See doc/decisions/2_in_progress/store_name.md

atmaxinger replied:
> Please be very faithful to the public API exactly as it is. 

Yes, I'm trying. but the current implementation (meaning the implementation before COW) also is not faithful. It only has two cases:

1. The `key` is `NULL` -> returns 0
2. Everything else -> returns what is stored as unescaped name, event if it is `NULL`.

It does never explicitly check whether the name itself is `NULL` to return an empty string.

atmaxinger replied:
I have updated `keyName` and `keyUnescapedName` now to conform to the API description

atmaxinger replied:
Okay this creates too much trouble. The implicit assumption in many function is, that a `NULL` keyname returns `NULL`. Reverting it back.

kodebach replied:
A bit of explanation: There was no `if` check for `key->key == NULL`, because that can never happen, when a key is created properly via the API. However, adding an assert just in case would make sense. The same question probably transfers to the new COW struct.

There was no `return ""` and there should never be one, because that is **not** a valid keyname.

atmaxinger replied:
Yes this can actuall happen! 

`ksLookupByName` creates its own key without the public API. If the `name` passed in is not a valid key name, it will be `NULL` as the return value of `keySetName` is never checked ....

https://github.com/ElektraInitiative/libelektra/blob/e9f2cab6f743f63746bfb800a3b4d14a29fe7df6/src/libs/elektra/keyset.c#L2608-L2627

atmaxinger replied:
Well this was apparently the root of the problem. Fixed this, now everything works again.

kodebach replied:
> `ksLookupByName` creates its own key without the public API

Then this is a problem in `ksLookupName`. In `keyName` we only need to expect (`if` check) what is possible via the public API, but we should still guard against (`assert`) other issues.

In the specific case, `ksLookupByName` should return `NULL`, when `keySetName` fails. Anything else is simply a bug.

kodebach commented on src/libs/elektra/keyname.c:
```diff
@@ -508,25 +557,17 @@ ssize_t keySetName (Key * key, const char * newName)
-		key->keyUSize = 0;
-		clear_bit (key->flags, (keyflag_t) KEY_FLAG_MMAP_KEY);
-	}
+	keyDetachKeyName (key);
```
The full detach is not necessary here. Like the old code shows, we don't need to copy the existing name. We just need a pointer value that is valid for `realloc`. Setting the pointer to `NULL` fulfils this purpose.

Since `keySetName` is called _a lot_, we really should avoid allocations here as far as possible.

kodebach commented on src/libs/elektra/keyname.c:
```diff
@@ -592,28 +633,16 @@ ssize_t keyAddName (Key * key, const char * newName)
-		char * tmp = elektraMalloc (key->keySize);
-		memcpy (tmp, key->key, key->keySize);
-		key->key = tmp;
+	keyDetachKeyName (key);
```
Note: here we need to full copy, because we want to add to it.

kodebach commented on src/libs/elektra/keyname.c:
```diff
@@ -1485,71 +1502,58 @@ static size_t keyAddBaseNameInternal (Key * key, const char * baseName)
 		}
 	}
 
-	size_t newKeySize = key->keySize + escapedSize;
-	size_t newKeyUSize = key->keyUSize + unescapedSize;
-	if (test_bit (key->flags, KEY_FLAG_MMAP_KEY))
-	{
-		// key was in mmap region, clear flag and copy to malloced buffer
-		char * tmp = elektraMalloc (newKeySize);
-		memcpy (tmp, key->key, key->keySize);
-		key->key = tmp;
+	size_t newKeySize = key->keyName->keySize + escapedSize;
+	size_t newKeyUSize = key->keyName->keyUSize + unescapedSize;
 
-		tmp = elektraMalloc (newKeyUSize);
-		memcpy (tmp, key->ukey, key->keyUSize);
-		key->ukey = tmp;
+	keyDetachKeyName (key);
 
-		clear_bit (key->flags, (keyflag_t) KEY_FLAG_MMAP_KEY);
-	}
-	else
-	{
-		elektraRealloc ((void **) &key->key, newKeySize);
-		elektraRealloc ((void **) &key->ukey, newKeyUSize);
-	}
+	elektraRealloc ((void **) &key->keyName->key, newKeySize);
+	elektraRealloc ((void **) &key->keyName->ukey, newKeyUSize);
```
This could also be combined probably. Maybe the solution in general, is adding parameters to `keyDetachKeyName` that indicate the new sizes after detaching. Then wen can realloc in one go.

markus2330 replied:
If this gets changed anyway, we probably do not optimize it here?

kodebach replied:
What do you mean?

What I wanted to say is the `keyDetachKeyName` makes a copy, which is then immediately passed to `elektraRealloc` to resize it.  `keyDetachKeyName` could allocate the right size immediately and save one reallocation (and potential copy).

atmaxinger replied:
done.

markus2330 replied:
Commenting creates notifications for me. So unless this is wanted, a simple thumbs-up or "Resolve conversation" is preferred.

kodebach commented on tests/abi/testabi_ks.c:
```diff
@@ -221,8 +221,10 @@ static void test_ksReference (void)
-	succeed_if (keyGetRef (k2) == 2, "reference counter after ksdup");
+	// COW - key references stay the same
+	succeed_if (keyGetRef (k1) == 1, "reference counter after duplication of keyset");
+	succeed_if (keyGetRef (k2) == 1, "reference counter after ksdup");
```
You should add tests for when the references change, i.e. after `ks1` has been modified.

kodebach commented on src/libs/elektra/keyset.c:
```diff
@@ -235,6 +290,7 @@ KeySet * ksVNew (size_t alloc, va_list va)
 	KeySet * keyset = 0;
 
 	keyset = (KeySet *) elektraMalloc (sizeof (KeySet));
+	memset (keyset, 0, sizeof (KeySet));
```
Technically, I believe the C standard doesn't mandate `NULL == 0`. So `memset` to zero might not correctly set `NULL` pointers.

atmaxinger replied:
See 6.3.2.3 Pointers, paragraph 3.

https://www.open-std.org/JTC1/SC22/WG14/www/docs/n1124.pdf

_An integer constant expression with the value 0, or such an expression cast to type `void *`, is called a null pointer constant._

atmaxinger replied:
You're technically right in that it doesn't mandate the same for `char *`.

kodebach replied:
That's what I meant: https://stackoverflow.com/a/9894047. See also: https://stackoverflow.com/a/61184560.

It is a super technical detail that probably will never have any impact in practice, so I'd just leave it as is.

mpranj replied:
Also, why not use calloc instead of malloc+memset?
https://stackoverflow.com/questions/2688466/why-mallocmemset-is-slower-than-calloc

atmaxinger replied:
Probably because the `elektraMalloc` code was already there... The new functions already use `elektraCalloc`. Will change this here too.

kodebach commented on src/libs/elektra/keyset.c:
```diff
@@ -1099,12 +1172,15 @@ ssize_t ksRename (KeySet * ks, const Key * root, const Key * newRoot)
 	if (keyGetNamespace (root) == KEY_NS_CASCADING || keyGetNamespace (newRoot) == KEY_NS_CASCADING) return -1;
 
+	// TODO: do we really need to detach here?
+	keySetDetachData (ks);
```
Yes, we need that. The function changes the keys contained in this keyset **without** changing them in other keysets they might be in. That's why `ksRenameInternal` has the "dup in-place".

This function is a bit weird and could be explained better. What it does is essentially the following, but much more optimized.

```c
elektraCursor end;
for (elektraCursor it = ksFindHierarchy (ks, root, &end); it < end; it++) {
    Key * k = elektraKsPopAtCursor (ks, it);
    Key * dup = keyDup (k); // needed because KEY_FLAG_RO_NAME
    keyDel (k);
    keyReplacePrefix (dup, root, newRoot);
    ksAppendKey (ks, dup);
}
```

But with all the COW, there may now be a better way to achieve this.

kodebach commented on src/libs/elektra/keyset.c:
```diff
@@ -1213,29 +1291,43 @@ ssize_t ksCopyInternal (KeySet * ks, size_t to, size_t from)
 	{
-		if (root->keyUSize == 3)
-		{
+		struct _KeyName * copy = keyNameCopy (root->keyName);
+		struct _KeyName * old = root->keyName;
+		((Key *) root)->keyName = copy;
+		keyNameRefInc (copy);
```
Not needed. Should have been marked as a HACK as well, but what we do here is completely undone before we return. Since we assume a single threaded context, this is okay, because nobody but us sees the modified version.

It's the same as in `ksBelow` where you added "DANGER".

atmaxinger replied:
Oh I ensure you this is needed ;). This was I bug that I spent way too much time debugging. If a key is in the keyset that is a duplication of `root` this completely breaks. There are some test cases that do exactly that.

kodebach replied:
Hm, okay that would be an issue. But then the detaching should only happen for that one case. If the keys don't share a name, we should not unnecessarily detach since that is a permanent operation.

kodebach replied:
Also couldn't the same issue happen in all the other functions that use this hack? If so, we need to fix those too.

atmaxinger replied:
We don't detach here. We just temporarily git the `root` key its own copied name. It gets it old name back a few lines after.

atmaxinger replied:
The other functions don't need it as this behaviour is special to `ksSearchInternal`.

atmaxinger replied:
Or at least there isn't a single test case that shows this behaviour in other functions.

kodebach replied:
> We don't detach here

Oh yeah, I misremembered. I'll have to look at this and the other functions again... Still seems a bit fishy to me that things are different here

kodebach replied:
Okay, so a few things I found:

1. If `root` itself is in `ks`, I think this code is just broken (even before COW). The `ukey` changes below would always change the name of a key that is in `ks` and therefore skew the result of `ksSearchInternal`.
2. The COW changes, extend this to any COW key that share a name struct with `root`.

In those two cases, do actually need to copy the name. In all other cases, it should be safe to change `ukey` temporarily. Luckily, the two cases are easy to detect, because we already searched for a key in `ks` with the same name as `root` above. Iff `search > 0` then we need the copy.

The exact same logic (1 & 2) applies to `ksBelow` and `ksCut`, but only when `root` uses the cascading namespace. In those functions it's not so easy to detect the two cases. However, we can just make the copy only if `root` is cascading. However, please add a `TODO`, since we plan to prohibit inserting cascading keys into keysets. When that is done, the copy can be removed again.

AFAICT there are no test cases for all that (probably nobody thought of these edge cases before), but it should be easy to add some and confirm my findings.

atmaxinger replied:
Okay, I've added some tests (the first 12 in `test_ks_cow.c`) for `ksBelow`, `ksCut` and `ksFindHierarchy` for those scenarios:

1. `root` itself is in `ks`
2. A copy of `root` is in `ks`
3. A cascading `root` is in `ks`
4. A copy of a cascading `root` is in `ks`


`ksBelow` has issues with (3), funnywise not with (4).
`ksCut` has no issues.
`ksFindHierarchy` has issues with (3) and (4)

kodebach replied:
Okay so `ksCut` doesn't use `ksFindHierarchy` (even though it could) and uses `elektraKsFindCutpoint` instead. The difference is that `ksFindHierachy` uses `ksSearchInternal` twice, while `elektraKsFindCutpoint` uses only one `ksSearchInternal` and then searches linearly from there. We should definitely update that, because doing two `O(log n)` binary searches is definitely better than one `O(log n)` binary search and then a `O(n)` linear search, even if we sometimes need to make a name copy.

That said, I believe there is still an issue. Try adding e.g. `keyNew ("/root/abc", KEY_END)` to the keyset in the tests. AFAICT this key will never be removed by `ksCut` (or returned by `ksBelow`), which is inconsistent with the fact that `/root` **is** removed. So either we need to also remove cascading keys below `root` or should leave `root` untouched too. I guess this, was always a bug in `ksCut` and is just one more reason to prohibit cascading keys in keysets. But that's a bigger task, so better fix it for now by making a name copy.

atmaxinger replied:
Okay with the updated tests only the `ksBelow` issues remain.
The `ksFindHierarchy` issues were caused by a misunderstanding of mine of how hierarchies work.

kodebach replied:
Okay so after a lot of confusion and mentally stepping through `ksCut` again and again, I now know why `ksCut` isn't broken. It's a total coincidence. It works, because `KEY_NS_CASCADING` is the first namespace in the order and because it searches linearly from the `ksSearchInternal` result. If it wasn't `ksCut` would return cascading keys for any `root` with a namespace before `KEY_NS_CASCADING`, because of the weird way that `keyIsBelowOrSame` works.

So, two options:
1. Leave `ksCut` as is and add two code comments to the `KEY_NS_CASCADING` section:
     1. `// NOTE: Works only because `KEY_NS_CASCADING` is the first namespace`
     2. `// TODO: Should use ksFindHierarchy like ksBelow`
2. Update `ksCut` to use `ksFindHierarchy` and implement the same fix  as for `ksBelow`

In any case, `ksFindHierarchy` should only do a `keyNameCopy` when actually required, i.e. `if (search > 0)`.

atmaxinger replied:
>  i.e. if (search > 0).

It's `search >= 0` btw 😉

markus2330 replied:
(Didn't read everything in the thread)

> I guess this, was always a bug in ksCut and is just one more reason to prohibit cascading keys in keysets.

Yes, we finally need a decision that KeySets are only allowed to have only keys of the same type. I added to #4645. I think this can directly go to "Decided" state, I don't think there are any argument for allowing such mixed KeySets.

> But that's a bigger task, so better fix it for now by making a name copy.

If there is a decision that we change something, we don't need to fix a bug that gets fixed with implementing the decision anyway (at least during 0.9.*).

> The ksFindHierarchy issues were caused by a misunderstanding of mine of how hierarchies work.

To me it is also not clear how the function is supposed to work, the documentation is poor.

> I think fixing non-broken code is not within the scope of this PR

I fully agree.

> @markus2330 rewriting ksCut may be a topic for FLOSS?

Issues for FLOSS are more-or-less already all taken for this term. C issues in particular are not very popular anyway.

Furthermore, we first need to decide what to do with ksFindHierarchy vs. ksCut. I initially thought this doesn't need a decision but as we are again and again discussing it, we probably should summarize current knowledge about which function to put in the core and which in a helper lib. (Or to keep both if its simply not worth the effort to get rid of one of them.)

kodebach replied:
> (Didn't read everything in the thread)

If you look at the last few comments, you'll see the issue is fixed (also why the thread is "resolved").

A quick summary: We found a flaw in the logic that explains the hacky "just edit `ukey` directly" code in `ksCut`, `ksBelow` and `ksFindHierarchy`. Some of that flaw existed before COW, other parts only exist through COW. In `ksCut` this doesn't result in a bug, because it balances out with the fact that cascading keys are the first in the order of namespaces. In `ksBelow` and `ksFindHierarchy` the issue has been fixed.

The other part about "non-broken code" was about the fact that `ksCut` does not use `ksFindHierarchy` and instead searches linearly for the end of the cut.

> To me it is also not clear how the function is supposed to work, the documentation is poor.

The part that caused confusion was how `Key`s are ordered in a `KeySet` and what "below" means for cascading names. The docs have already been updated in this PR https://github.com/ElektraInitiative/libelektra/commit/e47951a7ec06bbc248f43e2cca01169edf3139a9

> Yes, we finally need a decision that KeySets are only allowed to have only keys of the same type.
>  we first need to decide what to do with ksFindHierarchy vs. ksCut

I'll create those decisions since they either are 100% API or closely related to API questions.

> which function to put in the core and which in a helper lib

Recently, I've come to believe that the core should use indexes (i.e. `elektraCursor`) more, e.g. `ksSearch` not `ksLookup`. Similarly, `ksCut`/`ksBelow` can be replaced with `ksFindHierarchy` and a multi-key equivalent of `ksAtCursor`. I'll go into detail in the decision, but there is already a hint of that idea in the `KeySet` use cases in #4697.

kodebach commented on src/libs/elektra/keyset.c:
```diff
@@ -1629,9 +1741,12 @@ while ((key = ksNext (ks))!=0) {}
+
+	keySetDetachData (ks);
 
-	ks->cursor = 0;
-	ks->current = 0;
+	ks->data->cursor = 0;
+	ks->data->current = 0;
```
1. The internal cursor will be removed soon. Ideally you'd remove it now, but I think there may still be some uses left.
2. Why is it in `ks->data` and not directly in the KeySet? Detaching for `ksRewind` and `ksNext` seems bad. They don't modify any data.

atmaxinger replied:
1. Yes, but most `ks` methods still use the iterators ...
2. Good point!

atmaxinger replied:
I put `cursor`and `current` back into the `KeySet`.

kodebach reviewed (changes requested):
LGTM, some feedback below

atmaxinger commented on src/include/kdbprivate.h:
```diff
@@ -235,6 +243,54 @@ struct _Key
+	uint16_t refs;
+};
+
+// private methods for COW keys
+struct _KeyName * keyNameNew (void);
+struct _KeyName * keyNameCopy (struct _KeyName * source);
+uint16_t keyNameRefInc (struct _KeyName * keyname);
+uint16_t keyNameRefDec (struct _KeyName * keyname);
+uint16_t keyNameRefDecAndDel (struct _KeyName * keyname, bool deleteData);
+void keyNameDel (struct _KeyName * keyname, bool deleteData);
+
+void keyDetachKeyName (Key * key);
+
+struct _KeyData * keyDataNew (void);
+uint16_t keyDataRefInc (struct _KeyData * keydata);
+uint16_t keyDataRefDec (struct _KeyData * keydata);
+uint16_t keyDataRefDecAndDel (struct _KeyData * keydata, bool deleteData);
+void keyDataDel (struct _KeyData * keydata, bool deleteData);
```
Haha yes I thought of this as well. But it was easier for the first major refactoring to keep the old member names. We can do a second refactoring round and consolidate the structs.

kodebach commented on src/libs/elektra/key.c:
```diff
@@ -423,43 +399,37 @@ Key * keyCopy (Key * dest, const Key * source, elektraCopyFlags flags)
 
 memerror:
-	elektraFree (dest->key);
-	elektraFree (dest->data.v);
+	keyNameRefDecAndDel (dest->keyName, !test_bit (dest->flags, KEY_FLAG_MMAP_KEY));
+	keyDataRefDecAndDel (dest->keyData, !test_bit (dest->flags, KEY_FLAG_MMAP_DATA));
 	ksDel (dest->meta);
```
We should document (and add to the release notes) that `keyCopy` now only allocates memory, if `KEY_CP_META` (or `KEY_CP_ALL`) is used. A `keyDup` of course still allocates memory (because it calls `keyNew`), but still knowing that `keyCopy` can work without allocations can be useful, e.g. reusing a `Key` for `keyCopy` instead of doing many `keyDup` has now much more effect.

atmaxinger replied:
Speaking of `keyDup`... currently, it is implemented as a `static inline` function in the header. This is fine, I guess. However, now with COW, we could make it more efficient. 

Currently, it always allocates a name (because of the public API of `keyName` will always allocate a name). With COW we don't need to do that anymore. But that would require converting it to a normal function and not `static inline`.

kodebach replied:
Hm, yeah it could be better. But I'd leave it for now. I believe the unnecessary allocation should be solved via a more generally useful optimization and not something specific to `keyDup`.

kodebach commented on tests/ctest/test_keyset_cow.c:
```diff
@@ -0,0 +1,349 @@
+
+	// Assert
+	// keys in the keyset are ordered alphabetically.
+	// - system:/other
+	// - /root
+	// - system:/root/abc
```
That order would be wrong. `/root` should be first. Keys are ordered by namespace according to 
https://github.com/ElektraInitiative/libelektra/blob/4283744cc4689792a8143d55f5f49e6dc1eac81c/src/include/kdb.h.in#L86-L96

atmaxinger replied:
Okay, but does that mean that a hierarchy, as defined by `ksFindHierarchy`, is defined by the namespace of the keys?

So 

- `system:/a`
- `system:/a/b`

are a different hierarchy from

- `user:/a`
- `user:/a/b`

?

kodebach replied:
Yes, the range returned by `ksFindHierarchy` will always contain keys of a single namespace only and that namespace must be the same as the one used by `root`.

It's obvious for the normal namespaces `system:/`, `user:/` etc., because by any definition `user:/` key is never below a `system:/` key or vice versa. But even if `root` uses `KEY_NS_CASCADING`, only cascading keys will be selected, so `ksFindHierarchy` uses a stricter definition of "below" than e.g. `keyIsBelow`.

If you think about it, there is no other way of doing this. Because we return only a `start` and `end` index, we can only select a continuous range. How would that work with `root = /a` for e.g.

```
user:/a/b
user:/a/b/c
user:/b/c
system:/a/b
```

If you want `user:/a/b` and `system:/a/b` to be selected, you have to select `user:/b/c` as well, because of the order.

We could change the order to sort by namespace after the rest of the name, i.e.:

```
user:/a/b
system:/a/b
user:/a/b/c
user:/b/c
```

But this now has a problem with `root = user:/a` where it is very obvious what that the result should always be:

```
user:/a/b
user:/a/b/c
```

And just to clarify, you can't sort `user:/a/b/c` before `system:/a/b`, because then we are back to the first problem, but now with `root = /a/b`

```
user:/a/b
user:/a/b/c
system:/a/b
user:/b/c
```

> **Note** The true reason is actually just that the namespace is stored as the first byte of the unescaped name and the order of keys is determined by a simple `memcmp` on the whole unescaped name. But as you can see, the current behaviour is still the best option, even if we'd use another way to sort keys.

atmaxinger replied:
Thanks for the detailed writeup! Will add some of it to the documentation of `ksFindHierarchy`.

kodebach commented on tests/ctest/test_keyset_cow.c:
```diff
@@ -0,0 +1,349 @@
+
+	// Assert
+	succeed_if_fmt(ksGetSize (below) == 2, "should contain 2 keys, was %zd", ksGetSize (below));
+	succeed_if (ksLookupByName (below, "/root", 0) != NULL, "should contain root");
```
As stated in https://github.com/ElektraInitiative/libelektra/pull/4691#discussion_r1027871203, either this key shouldn't be there, or `/root/abc` should also be there, if you add it to `ks`. Please adapt this one and the other tests.

atmaxinger replied:
All tests or only the tests with cascading keys?

kodebach replied:
At least to the ones with a cascading `root`. For the other's it doesn't really matter, but it also wouldn't hurt.

atmaxinger replied:
Done, added it to all testcases

kodebach commented on src/libs/elektra/keyset.c:
```diff
@@ -1269,7 +1269,9 @@ ssize_t ksCopyInternal (KeySet * ks, size_t to, size_t from)
- * Searches for the start and optionally end of the key hierachy rooted at @p root in @p ks.
+ * Searches for the start and optionally end of the key hierarchy rooted at @p root in @p ks.
+ * The hierarchy will only contain keys in the same namespace as @p root.
+ * If @p root is a cascading key, only cascading keys will be part of the hierarchy.
```
```suggestion
```

Since we plan to disallow adding cascading keys to keysets, that sentence will just be a future inconsistency. The sentence above it should also already make it clear that this happens, e.g. `KEY_NS_CASCADING != KEY_NS_SYSTEM`.

atmaxinger replied:
Yes and no. Other functions, like `ksBelow`, `ksCut`, ... treat cascading `root`s differently. This is also why I was confused what `ksFindHierarchy` does with cascading keys. I did not consider that `KEY_NS_CASCADING` would be treated like a "normal" namespace here.

atmaxinger replied:
So even if we disallow cascading keys in keysets in the future, the documentation still needs to be updated, as `ksFindHierarchy` will not do anything with them.

kodebach replied:
Well it's your decision. If you keep the note about cascading keys, please add a TODO (outside the doxygen comment) to remind us that this should be updated, when cascading keys are forbidden in keysets.

kodebach reviewed (approved):
LGTM. We need to find a way to make the build server happy (either fix `mmapstorage` in this PR, or fully disable its tests), but IMO this could be merged.

I just have one comment (https://github.com/ElektraInitiative/libelektra/pull/4691#discussion_r1027077773) where a tiny improvement could be made.

atmaxinger commented:
There is one thing we need to discuss for `mmapstorage` - the flags. I already had to add flags to `_KeySetData` because of the hashmap. So from my POV we could also add the flags to `_KeyName` and `_KeyData` aswell. 

If possible, we could make it `uint16_t` so that it does not use any more memory.

Adding the mmap flags to `_KeyName` and `_KeyData` would also cleanup the internal API a bit, because the `*Delete` and  `RefDecAndDelete` method's wouldn't need the `deleteData` flag anymore.

mpranj commented:
> So from my POV we could also add the flags to _KeyName and _KeyData aswell.

For `mmapstorage` I would need some "mmap flag" for each structure that is persisted in the storage. This is the way it was done before and it would use up only one bit of payload (similar to what the mmap flags in ksflag_t do).

From the current implementation I can only say that adding another uint16_t to _KeyName and _KeyData will not change the size of the structures at all (for 32 and 64bit machines).

> AFAIK we currently use an int, so while it's technically not clear, let's say 32 bits.

Maybe we can reduce this while we're at it. Do we really need 32 flags? 😄

kodebach commented:
> Do we really need 32 flags?

If we have more than one flags field probably not. Right now, with a single field, 16 bits could have been full at some point, since many flags need separate variants for (some of) the key itself, its name, value and metadata.

With separate fields, we'd effectively have 16 bits for key, name, value and metadata each. That'll be more than enough.

kodebach commented:
You could group name and value together, i.e. either both are mmap or both are not. It would be unnecessary to realloc the data/name if the other one changed, but it would work.

However, you cannot group them with the struct itself. Let's say you want to change the name, so you call `keySetName (key, ...)`. The `keySetName` function could know detect that the name and `key` itself are in mmap. But it cannot change that, because just receives a pointer from the caller. You can't change the pointer the caller is holder and therefore can't change where it points. It either points into the mmap region or it does not.

atmaxinger commented:
No, I'm not talking about grouping keyname and keyvalue together. This is what I have in mind and what is implemented:

- `struct _Key` needs it's own flag.
- `struct _KeyName` and its value(s) share a single flag (either both are in mmap or not)
- `struct _KeyData` and its value share a single flag  (either both are in mmap or not)

But I think we are on the same page on this one.

kodebach commented:
Oh yeah that should work fine, because `struct _KeyName` is completely private. There we can change the pointer when we need to, because the caller always passes us a `Key *` and never a `KeyName *`.

atmaxinger commented:
@mpranj I've implemented the flags now, and even added some convenience functions for retrieving and setting the flags on the internal structs for you 😉.

So the ball is in your park now. You can either work directly on my branch, or we disable the failing tests, merge and you fix mmapstorage and reenable the tests in a new branch.

atmaxinger commented:
@markus2330 @kodebach there was some talk about moving implementation details from the decision to `doc/contrib` in #4678. Is `doc/contrib` really the right folder for this? Wouldn't `doc/dev` be more appropriate?

kodebach commented:
Depends on the documentation, IMO. The parts that are relevant to developers using Elektra should be documented in `doc/dev` oder the doxygen API docs.

But internal details that will only ever be visible to people contributing to Elektra, should be in `doc/contrib`. That would probably apply to most of the COW stuff since the whole internal architecture is private. On the public side people only need to know when we do or don't allocate new memory IMO.

markus2330 commented:
Yes, `doc/contrib` is the right place for nearly everything related to the cache, as internals are only relevant for contributors.

`doc/dev` would be for people using Elektra, so they only need to know that there is a cache, where the files are and how to enable/disable the cache.

It is a good time now to suggest better names instead of contrib and dev. (In #4501)

atmaxinger commented:
Okay, I created a documentation for the copy-on-write stuff in `doc/contrib/copy_oon_write.md` now. I'm not too sure what the correct level of detail is, so reviews are always welcome 😉.

I have also extracted some information about `mmapstorage` into `doc/contrib/mmapstorage.md`. But I'm sure @mpranj would be much more qualified to write about it than me.

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+The main principle for the copy-on-write approach is simple: seperate the "real" data from the user-facing identifying entity, in our case `Key` and `KeySet`.
+Developers using Elektra only get references to `Key` and `KeySet` objects.
+Those objects themselves do not hold any data.
+They only contain references to other entities holding the data.
```
```suggestion
They only contain references to other entities holding the data.
This allows transparently sharing the data-holding entities between multiple identifying entities.
```

kodebach commented on doc/contrib/copy_on_write.md:
````diff
@@ -0,0 +1,85 @@
+
+The main entities for copy-on-write are depicted in the following diagram:
+
+```mermaid
+classDiagram
+    Key *-- "0..1" KeyName : keyName
+    Key *-- "0..1" KeyData : keyData
+
+    KeySet *-- "0..1" KeySetData : data
+
+    Key o-- "0..1" KeySet : meta
+    KeySetData o-- "*" Key : array
+
+    Key : - uint16_t refs
+    Key : - keyflag_t flags
+
+    KeyName : - uint16_t refs
+    KeyName : - uint16_t flags
+
+    KeyData : - uint16_t refs
+    KeyData : - uint16_t flags
+
+    KeySet : - uint16_t refs
+    KeySet : - ksflag_t flags
+
+    KeySetData : - uint16_t refs
+    KeySetData : - uint16_t flags
+```
````
The Diagram should probably have fields for the actual name, value and key array in the respective structs to show what is stored where.

atmaxinger replied:
If you look at the rendered diagram, the field names are on the relationships.

kodebach replied:
I saw that. I didn't mean the name of the `KeyName` field in `Key`. I meant the `char * name` field in `KeyName`. It's a bit weird that it is absent, when both `refs` and `flags` are shown.

atmaxinger replied:
Oooh.... you're right. I added them in `KeyName` and `KeyData`. However, the the key array `KeySetData` is represented by a relationship.

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+- `KeySet`: logically represents a collection of keys. No real data is actually stored in it.
+  - `KeySetData`: holds an array of `Key` objects. May be shared between multiple `KeySet` objects.
+
+When creating instances of `Key` and `KeySet` via `keyNew` or `ksNew`, we allocate the least amount.
```
```suggestion
When creating instances of `Key` and `KeySet` via `keyNew` or `ksNew`, we allocate the least amount possible.
```

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+  - `KeySetData`: holds an array of `Key` objects. May be shared between multiple `KeySet` objects.
+
+When creating instances of `Key` and `KeySet` via `keyNew` or `ksNew`, we allocate the least amount.
+For `Key`, we only allocate a `KeyName` entity.
```
With the current API I don't think even that would be necessary. Since `keyName` and `keyUnescapedName` only return const pointers, they could do:

```c
const char * keyName (Key * key) {
    // [...]
    if (key->keyName == NULL) return "/";
    // [...]
}

const void * keyUnescapedName (Key * key) {
    // [...]
    if (key->keyName == NULL) return "\x01\x00\x00";
    // [...]
}
```

But it might complicate things elsewhere (e.g. `ksCut`), so I propose:

```suggestion
For `Key`, we only allocate a `KeyName` entity (while not technically necessary, it significantly simplifies some internal uses).
```

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+For `KeySet`, we allocate no other entities.
+The missing entities will get allocated when needed.
+
+If the name of a key gets modified, we first ensure that the key is the singular "owner" of its `KeyName` entity.
+It it isn't, then a new `KeyName` object is allocated for it.
```
```suggestion
If the name of a key gets modified, we first ensure that the `Key` holds the only reference to its `KeyName`.
If it isn't, then a new `KeyName` object is allocated for it.
```

Since we're using reference counting it's easier to phrase it that way IMO. Otherwise, we'd have to define ownership semantics on top of the refcount.

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+If the name of a key gets modified, we first ensure that the key is the singular "owner" of its `KeyName` entity.
+It it isn't, then a new `KeyName` object is allocated for it.
+Depending on the type of modification, the value of the name is also copied.
+The modifications to the name are then performed on the now exclusively owned `KeyName` entity.
```
```suggestion
The modifications to the name are then performed on the new `KeyName` entity.
```

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+Depending on the type of modification, the value of the name is also copied.
+The modifications to the name are then performed on the now exclusively owned `KeyName` entity.
+
+The same applies for the value of a key, and for keysets.
```
```suggestion
The same applies for the value of a key, and for the data contained in keysets.
```

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+
+The same applies for the value of a key, and for keysets.
+
+When copying the name or value of a key, or data of a keyset, via `keyCopy`, `ksCopy`, `keyDup` or `ksDup`, point the according references to the entities of the `source` object.
```
```suggestion
When copying the name or value of a key, or data of a keyset, via `keyCopy`, `ksCopy`, `keyDup` or `ksDup`, we point the according references in the `destination` object to the entities of the `source` object.
```

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+
+When copying the name or value of a key, or data of a keyset, via `keyCopy`, `ksCopy`, `keyDup` or `ksDup`, point the according references to the entities of the `source` object.
+The reference counter of those entities is increased accordingly.
+The entities of the `destination` object will get their reference count decreased, and if they reach 0 they will be `free`d.
```
```suggestion
The previous entities of the `destination` object will get their reference count decreased, and if they reach 0 they will be `free`d.
```

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+The reference counter of those entities is increased accordingly.
+The entities of the `destination` object will get their reference count decreased, and if they reach 0 they will be `free`d.
+
+## Reference counting
```
I think this should be explained before the details above.

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+The reference count is stored in each entity in an unsigned 16-bit integer variable `refs`.
+The special value `UINT16_MAX` is reserved as an error indicator, so the maximum number of references for each entity is `UINT16_MAX - 1` or 65534.
+
+The reference counting mechanismns for `Key` and `KeySet` are part of the public API. The reference counting mechanismns for `KeyName`, `KeyData` and `KeySetData` are private.
```
```suggestion
The reference counting mechanismns for `Key` and `KeySet` are part of the public API.
```

AFAICT `KeyName`, `KeyData` and `KeySetData` are entirely private, so their refcount API obviously must be  private too.

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+
+The reference counting mechanismns for `Key` and `KeySet` are part of the public API. The reference counting mechanismns for `KeyName`, `KeyData` and `KeySetData` are private.
+
+A value of 0 for `KeyName`, `KeyData` and `KeySetData` is illegal, except directly after it is allocated or directly before it is going to be `free`d.
```
```suggestion
A reference count of 0 for `KeyName`, `KeyData` and `KeySetData` is illegal outside the public API function calls that cause the creation/deletion these structures.
```

"directly before/after" is very vague. The important part here is that effectively a struct is never stored with refcount 0. Either the refcount will increase before we return from a public API function, or the struct will be deleted. Maybe you have a better way to phrase this...

kodebach reviewed (approved):
LGTM, just some textual wording things

markus2330 commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+For `KeySet`, we allocate no other entities.
+The missing entities will get allocated when needed.
+
+If the name of a key gets modified, we first ensure that the key is the singular "owner" of its `KeyName` entity.
```
Please use terminology properly (somewhere write what owner is, what owns mean etc.). We can use owner explicitly for COW, it is currently not used anywhere else (we had another owner concept in the past, but this, except of small left-overs waiting to be removed, not present anymore).

```suggestion
If the name of a key gets modified, we first ensure that the key is the single owner of its `KeyName` entity.
```

atmaxinger replied:
This has already been changed. I think you commented on a pretty outdated version of the documentation.

markus2330 commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+The reference count is stored in each entity in an unsigned 16-bit integer variable `refs`.
+The special value `UINT16_MAX` is reserved as an error indicator, so the maximum number of references for each entity is `UINT16_MAX - 1` or 65534.
+
+The reference counting mechanismns for `Key` and `KeySet` are part of the public API. The reference counting mechanismns for `KeyName`, `KeyData` and `KeySetData` are private.
```
Do we have a use case for reference counting as part of the API?

Afaik keyGetRef is nowhere used. This should be part of references decision (@kodebach).

kodebach replied:
> Do we have a use case for reference counting as part of the API?

Bindings

> Afaik keyGetRef is nowhere used.

That part may be unnecessary.

> This should be part of references decision

Agreed, let's not discuss it here and leave things as they are.

markus2330 commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+
+## Integration with `mmapstorage`
+
+The `mmapstorage` plugin needs a flag for every entity to indicate whether the entity is stored via `mmap` or normally allocated via `malloc`. If an entity is flagged as being stored in `mmap`, it must **never** be freed using `free`.
```
Shouldn't this be in the mmap internals docu?

atmaxinger replied:
Well it plays a part in the implementation for COW also, so I think it fits in there.

markus2330 commented on doc/contrib/mmapstorage.md:
```diff
@@ -0,0 +1,18 @@
+
+The most important constraint for `mmapstorage` is that any structure (or bytes) that is an allocation unit (e.g. we `malloc()` the bytes needed for `KeySet` struct, so this is an unit) needs to have a flag to determine whether those bytes are actually `malloc()`ed or they are `mmap()`ed.
+
+The `mmapstorage` plugin only calls `munmap` in some error cases, so basically `munmap` is almost never done and the keyset is never invalidated.
```
```suggestion
The `mmapstorage` plugin only calls `munmap` in some error cases, so basically `munmap` is never done and the keyset is never invalidated.
```

kodebach replied:
```suggestion
The `mmapstorage` plugin only calls `munmap` in some error cases, so basically the returned keyset is never invalidated and `munmap` is never called for its data.
```

If you want it to be correct...

markus2330 commented on doc/contrib/mmapstorage.md:
```diff
@@ -0,0 +1,18 @@
+
+The `mmapstorage` plugin only calls `munmap` in some error cases, so basically `munmap` is almost never done and the keyset is never invalidated.
+
+During `kdbSet` the storage plugins always write to a temp file, due to how the resolver works.
```
Is this information related to the cache or the mmap storage? @mpranj

markus2330 commented on doc/contrib/mmapstorage.md:
```diff
@@ -0,0 +1,18 @@
+# mmapstorage
```
Isn't here also information about the cache? Maybe call the file simply `mmap`.

kodebach replied:
IMO the cache is different thing that needs separate docs. The current cache `cache` uses `mmapstorage`, but you can use `mmapstorage` standalone and that has all the same intricacies explained here.

markus2330 commented on doc/contrib/mmapstorage.md:
```diff
@@ -0,0 +1,18 @@
+Therefore the only case where we return a `mmap()`ed KeySet should be in `kdbGet`.
+
+When the `mmapstorage` was designed/implemented, not all structures had refcounters, so there was no way to know when a `munmap` is safe.
+This was simply out of scope at that point in time.
```
Actually we could now look at the ref counter of the KeySet, can't we? @mpranj

kodebach replied:
Doesn't really help, unless there is a callback into `mmapstorage` when `ksDel` is called or the refcount changes. `mmapstorage` could check the refcount. But it would only know whether at this moment `munmap` can be called. If it can't `munmap`, nothing changes, the data is still effectively leaked.

Also this would require that callers of `kdbGet` do a `ksIncRef`. Because right now, unless you use a binding the refcount of the `KeySet*` will very likely be 0.

kodebach replied:
However, in the refcount decision there are some ideas to make this work better.

mpranj replied:
Yes, with the refcounter we could know when munmap is safe, but we would need to iterate over all keys and data which is/was in the region.
E.g. some random key could still be used elsewhere possibly, in a different keyset, couldn't it?

Also yes, we would need a callback so we can munmap. We also need the original mmap region pointer to munmap it.

Also, the global keyset may be in that region, so again we would have to check if it is safe to munmap for all parts. So the solution won't be only a few lines of code I suspect.

markus2330 commented on doc/news/_preparation_next_release.md:
```diff
@@ -73,6 +73,13 @@ Take a look at the [new docs](../dev/mountpoints.md), if you need to know detail
+Thanks to _(Maximilian Irlinger @atmaxinger)_ our `Key` and `KeySet` datastructures are now fully copy-on-write!
+This means noticeably reduced memory usage for cases where keys and keysets are copied and/or duplicated!
+
+<!-- TODO: add some benchmarks -->
```
Looking forward to them :heart:

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -210,6 +220,23 @@ struct _Key
 	 */
 	size_t dataSize;
```
Also should be, imho, valueSize.

kodebach replied:
Again, was already called this (isn't not even a changed line). Also depending on the decision in #4683 may not even matter. Finally, it may be possible (but definitely more difficult) to use `tree-surgeon` here as well.

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -210,6 +220,23 @@ struct _Key
+	 */
+	uint16_t refs;
+
+	uint16_t flags;
```
We can specify here how many bits we actually need.

atmaxinger replied:
What do you mean? We already specify that it's 16 bits?

markus2330 replied:
We only need 3 bits or so.

```suggestion
	uint16_t flags : 3;
```

kodebach replied:
AFAICT the bitfields in C are a total mess. From what I found online, the actual memory layout of bit fields is entirely implementation defined. So adding the `: 3` may make the struct incompatible between different compilers in theory.

atmaxinger replied:
Okay but if we go this route, I'd rather split up each flag as a bitfield of 1.

```suggestion
	bool isInMmap : 1;
	bool anotherFlag : 1;
	bool yetAnotherFlag : 1;
```

Would improved readability of the code IMHO.

markus2330 replied:
Great suggestion. Doesn't need to be within this PR, unless you touch this anyway.

atmaxinger replied:
> Doesn't need to be within this PR, unless you touch this anyway.

Not much of a problem for me, they have already been touched so I can touch them again, at least for the COW structures. For `Key` and `KeySet` I need to look how many tests depend on `flags` field being there.

> may make the struct incompatible between different compilers in theory

Yes, but is this really a concern? We're talking about purely internal datastructures here. Is it likely, that parts of Elektra will be built using different compilers?

kodebach replied:
I looked into it a bit more and while the C standard seems to have only very vague definitions of bitfields, the AMD64 ABI is more specific. So I think we should do

```suggestion
	bool isInMmap : 1;
	int : 15;
```

The unnamed `int : 15` field reserves 15 more bits for future use. We don't need to think about the type and can just use a simple `int`, because the AMD64 ABI says: "Unnamed bit-fields’ types do not affect the alignment of a structure or union."

Adding more flags should be a completely transparent change (even for `mmap`) by just adding another `bool` and reduce the size of the unnamed field.

I also found the following infos, which all point toward `bool flag : 1` being better than `uint16_t flags : 3` in general:
- The C standard only explicitly allows bitfields with types `bool`, `signed int` and `unsigned int`. For just `int` it is implementation defined, whether the field is signed or unsigned. Using any other types would be entirely implementation defined.
- The memory layout is actually not entirely implementation defined. The C standard says that adjacent bit-fields in a struct should be packed together. But it is implementation defined, whether a bit-field can overlap two storage units. So by only using bit-fields of size 1, we are definitely covered by the standard and everything must be packed in the minimum possible space.

kodebach replied:
> Yes, but is this really a concern? We're talking about purely internal datastructures here. Is it likely, that parts of Elektra will be built using different compilers?

See my previous comment. Using `bool` should be fine on every compiler.

I wasn't super concerned about the issue anyway. I just wanted to mention that bit-fields are not fully specified in the standard, since normally try to avoid implementation defined behaviour.

atmaxinger replied:
Done.

markus2330 commented on src/include/kdbprivate.h:
```diff
@@ -235,6 +262,89 @@ struct _Key
+	uint16_t flags;
+};
+
+// private methods for COW keys
```
Can you put all the stuff here in a separate header file? (See doc/decisions/3_decided/header_include.md)

kodebach replied:
> See doc/decisions/3_decided/header_include.md

That's not even close to being implemented. I don't see any benefit to moving things. We can't even move them to the correct location, because the folder structure is still all wrong...

atmaxinger replied:
Okay, so we keep it in `kdbprivate.h` for the time being?

markus2330 reviewed (changes requested):
Fascinating work! I especially like the many beautifully written tests!

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+
+- `Key`: logically represents a key. No real data is actually stored in it.
+  - `KeyName`: holds the name of a key. May be shared between multiple `Key` objects.
+  - `KeyData`: holds the value of a key. May be shared between multiple `Key` objects.
```
It's already called `KeyData` in the decision and I assume the name comes from the fact, that the field in `struct _Key` was already called `data`. So going to `value` would actually be a new name in some sense.

Let's leave it for now. If you want I can rename it later with `tree-surgeon`, one more renaming wouldn't change much.

atmaxinger replied:
Yes, @kodebach's reasoning is correct. I tried to name everything as close as possible to the existing terminology.

kodebach commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+- `KeySet`: logically represents a collection of keys. No real data is actually stored in it.
+  - `KeySetData`: holds an array of `Key` objects. May be shared between multiple `KeySet` objects.
+
+When creating instances of `Key` and `KeySet` via `keyNew` or `ksNew`, we allocate the least amount.
```
> The least amount is for me [...]

That's why it is explained afterwards...

> would that even be possible with the current implementation?

I made a similar comment (https://github.com/ElektraInitiative/libelektra/pull/4691#discussion_r1032767092) and I believe the text was changed already.

> This should be discussed in the decision.

atmaxinger replied:
> would that even be possible with the current implementation? 

Yes. Just set the initial ref count to anything > 0.

atmaxinger replied:
Alternatively, interpret a `NULL` for `keyName` as `/` in `keyGetName` et al.

kodebach commented:
@markus2330 I think you need to read the docs again. Not sure if it's just https://github.com/ElektraInitiative/libelektra/pull/4691/commits/dfcf0468844200756fe875489b593ac4d80281a6, but there definitely have been changes between the start of your review and when you posted it. Some of your comments are on already outdated lines. GitHub should give you a button somewhere to just show changes since you review...

atmaxinger commented on doc/contrib/copy_on_write.md:
```diff
@@ -0,0 +1,85 @@
+The two basic Elektra datastructures `Key` and `KeySet` implement full copy-on-write (COW) functionality.
+If a key or a keyset gets copied, only a shallow copy with references to the original name and value is created.
+Only if it gets modified the necessary memory is actually allocated.
+As a consequence, duplicated keys or keysets only require a fraction of the memory compared to their source counterparts.
```
Wouldn't `doc/dev/data-structures` be a better fit? In there we already describe `Key` and `KeySet`.

atmaxinger commented:
Seems like COW also decreases memory footprint not only in theory but only in practice. 😉 I used `valgrind --tool=massif` to profile memory usage, and compare between the `master` branch and this branch. I used the `large` benchmark.

It's `18.9 MiB` for the master vs `15.3 MiB` for the copy-on-write max memory usage implementation. 🤗🎉🚀 Also, it's interesting to see the different memory patterns.

![grafik](https://user-images.githubusercontent.com/5960633/204850980-d603707d-d98c-4063-8dd3-ee4f5fd484da.png)

markus2330 commented:
These are **great** news and a great foundation for change tracking! Please add this info&graph to the release notes!

kodebach commented:
> Please add this info&graph to the release notes!

Maybe we can produce a slightly nicer graph for the release notes. For example, it might be confusing that the colors don't mean the same things in both graphs. Also the chosen functions are a bit weird. For example the COW version lists both `ksNew` and `ksVNew`, when `ksNew` is essentially just a wrapper around `ksVNew`, and `keySetRaw` is listed twice.

Is there a way you can produce the graph with just the overall memory consumption, or with the same functions on both sides?

atmaxinger commented:
> These are great news and a great foundation for change tracking! Please add this info&graph to the release notes!

Yup, but I probably should create a 'real-world' benchmark first. In the current benchmarks, key values are never changed. And duplications of keys only happens internally (i.e. within `kdbGet` and `kdbSet`), not in 3rd party code.

> Maybe we can produce a slightly nicer graph for the release notes. 

Yes

> Is there a way you can produce the graph with just the overall memory consumption, or with the same functions on both sides? Or maybe even both master and COW in a single graph (just the total memory use), to make comparing easier?

I'll have to look. This is just the default graph that `massif-visualizer` presents. Maybe I could just parse the massif output manually and create graphs in R.

kodebach commented:
> Maybe I could just parse the massif output manually and create graphs in R.

`massif` is not a niche tool, I would assume there is an easier way to do it...

atmaxinger commented:
I have now added the benchmark results in tabular form into `_preperation_next_release.md`.

Overall, I think the benchmarks look very encouraging!

I'm however not sure about the runtime benchmarks. While the benchmarks themselves output some timings, I have found them too fine grained for this. Instead I used `perf` and just reported elapsed time. IMHO this is probably a more realistic metric for applications using Elektra.

kodebach commented on doc/news/_preparation_next_release.md:
```diff
@@ -73,6 +73,46 @@ Take a look at the [new docs](../dev/mountpoints.md), if you need to know detail
+One interesting observation is that `kdbget.c` and `kdbmodify.c` used exactly the same memory.
+This can most likely be explained by internal caching within the memory allocator of `glibc`.
+
+We also performed runtime tests on the same benchmarks using `perf stat --repeat 13`.
+
+| Benchmark       | Current Implementation | Deviation | Copy-on-Write | Deviation | Runtime Increase |
+| :-------------- | ---------------------: | --------: | ------------: | --------: | ---------------: |
+| `createkeys.c`  |             0.209572 s |    0.36 % |     0.21987 s |    0.77 % |            4.9 % |
+| `deepdup.c`     |              0.23025 s |    0.47 % |    0.231804 s |    0.32 % |            0.6 % |
+| `large.c`       |              1.14038 s |    0.21 % |     1.14837 s |    0.21 % |            0.7 % |
+| `kdb.c`         |               1.9270 s |    2.63 % |     1.93354 s |    0.17 % |            0.3 % |
+| `kdbget.c`      |             0.145663 s |    0.17 % |     0.15763 s |    0.70 % |            8.2 % |
+| `kdbmodify.c`   |             0.146506 s |    0.19 % |    0.156347 s |    0.15 % |            6.7 % |
+
+Overall, the runtime performance hit is less than 10%.
+The more a program does, the less the additional overhead of the copy-on-write algorithms matter.
```
I guess these benchmarks mainly serve to show that while memory usage has gone done quick a bit, the runtime is mostly unaffected.

One important thing I would also highlight is that `keyDup` has become quite a bit faster. You can see this by comparing the difference between `createkeys.c` and `deepdup.c`, since `deepdup.c` is just `createkeys.c` plus a round of `keyDup`s. In "Current Implementation" the difference is ~21ms in COW it is only ~12ms. And that is with the naive `deepdup` approach that doesn't pre-allocate the right size of `KeySet`.

Finally, I would call it "Old Implementation" instead of "Current implementation". At the time these release notes are released, COW will be the current implementation.

atmaxinger replied:
> I guess these benchmarks mainly serve to show that while memory usage has gone done quick a bit, the runtime is mostly unaffected.

Yes

> One important thing I would also highlight is that keyDup has become quite a bit faster. 

Nice catch!

> Finally, I would call it "Old Implementation" instead of "Current implementation".

Good point.

atmaxinger commented:
I have moved the `copy_on_write.md` decision into `5_implemented` now. I think this makes sense, because it is, well, implemented now.

@markus2330 @mpranj how should we continue here? I'd rather not have this PR open for too long. It has quite a number of changes and we should not let `master` deviate too much. 

The main reason why we can't merge right now is because of the failing build. This is mainly due to tests looking for the `cache` and `mmapstorage` plugins.

We have 2 options here:

1. we fix `mmapstorage` now - this depends entirely on @mpranj 's timeline. As he still hasn't left a review on here I guess he is pretty busy otherwise and doesn't have too much time on his hand fixing it right now.
2. we disable those tests and enable them when @mpranj has time to fix the `mmapstorage` plugin.

markus2330 commented:
To get this done as fast as possible:

- don't put the decision to 5_implemented now but with a later PR where also left-over documentation gets updated (then I can be less strict with documentation in this PR).
- disable the mmapstorage tests as we already discussed.

It would be great if @mpranj can also review, so that there are no surprises when he tries to fix mmapstorage.

mpranj commented:
>  It would be great if @mpranj can also review, so that there are no surprises when he tries to fix mmapstorage.

I will review it this evening. I would propose that I disable the tests and document it, so that I know how to re-enable everything.

Before I disable the tests it would be great to not rebase again s.t. the commit hash does not change and can be easily undone. Is a rebase or any further work needed here?

atmaxinger commented:
> don't put the decision to 5_implemented now 

Done, moved it to "in progress"

>  Is a rebase or any further work needed here?

There are no merge conflicts, so I don't think a rebase is required.

kodebach commented:
> Done, moved it to "in progress"

That makes zero sense IMO. It should at least be in `4_partially_implemented`. But I'll let @markus2330 decide, if that's appropriate.

mpranj commented on doc/news/_preparation_next_release.md:
```diff
@@ -73,6 +73,49 @@ Take a look at the [new docs](../dev/mountpoints.md), if you need to know detail
+Thanks to _(Maximilian Irlinger @atmaxinger)_ our `Key` and `KeySet` datastructures are now fully copy-on-write!
+This means noticeably reduced memory usage for cases where keys and keysets are copied and/or duplicated!
+
+We run some very promising benchmarks, each were performed with 400,000 keys.
```
```suggestion
We ran some very promising benchmarks, each were performed with 400,000 keys.
```

mpranj reviewed (approved):
First of all: **absolutely amazing work @atmaxinger!** 🚀 

I started making some comments, but noticed that it would actually be best if I fix the mmap related things (additional flags) myself. You prepared the bitfields and everything, it's basically irrelevant how many flags I will need, and is not relevant to this PR at all.

I propose that we just disable the mmap related tests and merge. @atmaxinger should I disable it or what do you propose?

atmaxinger commented on src/include/kdbprivate.h:
```diff
@@ -236,26 +180,193 @@ struct _Key
 
 	/**
-	 * Reserved for future use
+	 * Is this structure and its data stored in an mmap()ed memory area?
```
Ah yes, good catch.

atmaxinger commented on src/include/kdbprivate.h:
```diff
@@ -272,33 +383,33 @@ typedef struct _KeySet KeySet;
 
-	uint16_t reserved; /**< Reserved for future use */
+	/**
+	 * Is this structure and its data stored in an mmap()ed memory area?
```
```suggestion
	 * Is this structure stored in an mmap()ed memory area?
```

atmaxinger commented:
> @atmaxinger should I disable it or what do you propose?

Well I don't think you have write access to my private branch, so I think I should probably disable them

mpranj commented on src/include/kdbprivate.h:
```diff
@@ -210,6 +128,32 @@ struct _Key
+	uint16_t refs;
+
+	/**
+	 * Is this structure and its data stored in an mmap()ed memory area?
```
So if a key data is changed you malloc a completely new keydata struct and then you malloc the data * afterwards, instead of just adjusting the values in the keydata struct and doing one malloc?

atmaxinger replied:
Yes, if the flag `isInMmap` is set. You can see this logic in the `*Detach*` functions.

mpranj replied:
I guess this can be done ... but why not reuse the struct in the mmap region?
Are there any benefits, except that we have one `if`-statement less?

atmaxinger replied:
No, it just never crossed my mind that this could be done.

kodebach replied:
> Are there any benefits, except that we have one if-statement less?

I think we could do something like this later on (pseudo-code!):

```c
void keySetData (Key * key, void * data, size_t dataSize) {
    KeyData * keyData = malloc (sizeof(KeyData) + dataSize);
    keyData->data = keyData + 1; // pointer arithmetic in terms of sizeof(KeyData)
    keyData->dataSize = dataSize;
    memcpy (keyData->data, data, dataSize);
    key->keyData = keyData;
}
```

Using only one allocation could improve cache locality (especially for small `dataSize`). Although that would need benchmarking. In any case the option is there. If `keyData->data` can be malloc when `keyData` is mmap this would be harder to do.

mpranj replied:
> Using only one allocation could improve cache locality

Good idea, but wouldn't there be alignment problems? Isn't char * byte-aligned while the struct would be aligned as the most strict member is aligned, aka pointers, thus 4 or 8 bytes in most cases?

kodebach replied:
> Good idea, but wouldn't there be alignment problems?

Well first of all the alignment problems only come up, when you do stuff on the stack. `malloc` always uses the biggest possible alignment, because it doesn't know what alignment you want. So the `keyData` pointer is definitely aligned correctly.

Technically, there could be a problem here, because I used `void * data`. But with `char * data` it would definitely be okay, because `keyData->data` can't be misaligned for a `char *`. If we want to support bigger alignments, we'd have to insert a padding. I'm not gonna post the code here, but [here](https://git.uclibc.org/uClibc/tree/libc/stdlib/malloc/malloc.c) is a simple implementation of malloc. From there we could extract the part the ensures the alignment (just some macros that round the size to the correct alignment).

mpranj replied:
You're right, I might have missed it alltogether: if the struct comes first and is strictly aligned due to the pointer members, anything that comes after it will have the same nice alignment.

Sorry for the confusion. I had this mistaken, as in the mmap regions usually something else comes after the data again, which could then be misaligned due to `char *`. This is not the case here.

Were you planning to work on this, i.e. would you mind if I do some benchmarks regarding this approach?

kodebach replied:
No, no plans here, just an idea I had. Feel free to work on this. I'd love to see benchmarks, to know whether this actually has any benefit.

mpranj commented:
> Well I don't think you have write access to my private branch, so I think I should probably disable them

I can copy your changes to branch where I have access and then work on it. It would still show what you did and what I did and if I do not rebase, then this PR is automagically shown as merged too...
So it would not be a problem I think.

atmaxinger commented:
> I can copy your changes to branch where I have access and then work on it. 

Yeah, I think you can just "fork" my branch. Feel free to do so.

kodebach commented:
> Well I don't think you have write access to my private branch

@atmaxinger You should see an option like this at the top of the PR (in the sidebar on the right)

![Screenshot from 2022-12-02 19-25-00](https://user-images.githubusercontent.com/23529132/205360676-d2d78659-7aae-40b4-be3b-28ef29ac3105.png)

Checking the option should give anybody with write access to the main repo (including @mpranj) write access to only the branch this PR comes from.

atmaxinger commented:
@kodebach Yes! Nice it is actually enabled right now.

@mpranj Guess you can push to my branch directly.

atmaxinger commented:
@mpranj you also need to disable `testshell_markdown_mmapstorage_shelltests`

atmaxinger commented:
Seems like there are some minor memory leaks, I'll take a look

mpranj commented:
Thank you @atmaxinger for this amazing contribution and the intense work! 🥇 

Thanks to @kodebach and @markus2330 for the reviews and the detailed insights.
